### ISSUE TITLE
Support OneToOne relationships

### ISSUE THREAD

**Issue by jmmills** [josiahcarlson/rom#61]
There are cases when you might want to have a OneToOne relationship between objects rather than a OneToMany or ManyToOne

**Comment by josiahcarlson** [comment]
Aside from the concept of a "one to one" relationship, what functinoality did you want to be supported? Did you want bi-directional cascading deletes/restricts/no action? Did you want both sides of the OneToOne relationship to keep a reference to the other so that you can query against the object it references?

**Comment by jmmills** [comment]
I don't really have an opinion about cascading operations, although that would seem to be consistent with how other relations are done. My use case is specific to an application I'm working on that uses ROM for data persistence, but a more generically understood use case would be representing Tree data. For example:

    import rom
    
    class TreeNode
      parent = rom.OneToOne('TreeNode', required=False)
      left = rom.OneToOne('TreeNode', required=False)
      right = rom.OneToOne('TreeNode', required=False)
      name = rom.String()
      value = rom.String()
    
      def add_node(self, node, side)
        node.parent = self
        node.save()
    
        if side == 'left':
          self.left = node
        else:
          self.right = node
    
        self.save()
    
        return self
    
      def add_left(self, node):
        return self.add_node(node, 'left')
    
      def add_right(self, node):
        return self.add_node(node, 'right')
    
      def slow_search(self, name, side=None)
        pass

**Comment by josiahcarlson** [comment]
You can use a `rom.ManyToOne` column by itself wherever you wanted to use the (missing) `OneToOne`. Right now, it will work exactly the way you want, even if it isn't spelled the way you want.

I'll probably add a "real" OneToOne column soon (with cascading options), but try `rom.ManyToOne`.

**Comment by jmmills** [comment]
Is it a conditional relationship? MaybeHasMany? Sorry if my nomenclature seems silly.

**Comment by josiahcarlson** [comment]
Is what a conditional relationship?

In a `ManyToOne` relationship, because there are *many*, the *many* have the id reference to the *one*. That is what you want. Your `.parent`, `.left`, and `.right` are just pointers, and you don't need to have a reverse `OneToMany`. Just use a `ManyToOne` any time you need a pointer to something.

In the case of a tree, cascading deletes would be convenient if you could delete the root and have the entire tree delete itself, which is why I'll probably end up building a `OneToOne`, which will probably borrow much of the underlying code of `ManyToOne`, plus add cascading deletes. Who knows, this may even push me to build `ManyToMany` while I'm at it.

**Comment by jmmills** [comment]
Add column type inflation to the mix and you will have most of the features of Perl's DBIx::Class, but for Redis stored objects... 
I'm using this for a project at work, so I will do what I can to submit any patches that come up with (none as of yet).
Have you thought of supporting plugins?

**Comment by josiahcarlson** [comment]
I have no idea what "column type inflation" is, and Google is not helpful. Is it something Perl related? Can you provide a link or description of what you mean?

I haven't thought about adding plugins, because I don't know what someone might want in terms of functionality out of a plugin. What behaviors would a plugin change or add?

**Comment by jmmills** [comment]
Column inflation would be the ability serialize and deserialize column values to and from custom classes (i.e. types). Perhaps, I'm just missing how this is done with ROM.

It's probably just easiest to explain with example code:

    
    class KittenPaw:
    
      def __init__(self, paw):
        self._paw = paw
    
      def scratch(self):
        return self._paw
    
    class FluffyKitten(rom.Model):
      paw = rom.Integra(type=KittenPaw)
    
    
    kitten = FluffyKitten.get(id)
    kitten.paw.scratch()  
    kitten.paw = KittenPaw(4)
    kitten.save()
    

 As far as plugins, I was just thinking about any type one might want to change default behavior.

**Comment by josiahcarlson** [comment]
In most cases, you should be able to build your own column type to do exactly what you want, and it is pretty straightforward, even if it is mostly undocumented.

```
class KittenColumn(rom.Column):
    _allowed = KittenPaw
    def _to_redis(self, value):
        # turn the KittenPaw value into a 2.x string or 3.x bytes and return it
    def _from_redis(self, value):
        # value can be one of:
        # None, a KittenPaw object, or a serialized KittenPaw object returned from _to_redis
        # return None or the KittenPaw object
```

If you want to index this data, you would need to write your own keygen function.

Note that rom doesn't do anything special with mutable columns, so if your object is mutable, you will need to explicitly set the column value after changing your attribute. As in:

```
kitten = FluffyKitten.get(id)
kitten.paw._paw = 'hello'
kitten.paw = kitten.paw # necessary for the .save() to know the kitten.paw column changed
kitten.save()
```

**Comment by jmmills** [comment]
Can you do __setattr__ magic with mutable sub-classes of rom.Column?

    class KittenColumn(rom.Column):
        # boilerplate
        def __setattr__(self, *args, **kwargs)
             super(KittenColumn, self).__setattr__(*args, **kwargs)
             self.save() 


Of course in the real world you would want to make sure you are actually changing the value of the attribute before calling save.

**Comment by josiahcarlson** [comment]
That doesn't work because `KittenColumn` (and all other columns that derive from `rom.Column`) just implements the descriptor protocol (it is actually a lot like `property`). When you access `kitten.paw._paw`, you are actually referencing a `KittenPaw._paw`, because `kitten.paw` returns a `KittenPaw` (like a property would).

To remove the need for `kitten.paw = kitten.paw`, you would need to teach your `KittenPaw` object how to report back to the entity that it was changed. At present, the `Column._from_redis()` API that is available does not provide sufficient information to allow that reporting to take place. But even if it *did* provide sufficient information for that reporting to take place, your `KittenPaw` class would need to end up as something like...

```
class KittenPaw(object):
    def __init__(self, _paw, ent, attr):
        self.__dict__.update(locals())
        self.__dict__.pop('self')

    def __setattr__(self, attr, value):
        self.__dict__.update(**{attr:value})
        setattr(self.__dict__['ent'], self.__dict__['attr'], self)
        # calling model.save() should pick up these changes
```

Mutable columns are effectively impossible to handle automatically for arbitrary user-defined types, and can even be a chore when manually instrumenting them - anything that "modifies" your mutable object must report back to the containing entity that something changed.

For your mutable custom column, the simplest way to ensure your changes make it back into Redis is to just re-assign the column after changes to the column data. As in: `kitten.paw = kitten.paw` after altering data in the `kitten.paw`.

**Comment by josiahcarlson** [comment]
I just released version 0.31.0 that has support for `OneToOne` columns, as well as 2 more `on_delete` actions: `'set null'` and `'set default'` (`'no action'`, `'cascade'`, and `'restrict'` were already available).
